Fix -DHUNTER_ENABLED=0 builds

ringerc opened the pull request:
Fixes #1

ringerc committed:
Case of Find modules didn't match CMakeLists.txt usage

There were a number of case sensitivity issues in the CMake files relating
to differing case between the Hunter files and the ones in cmake/ (#1)

Also fixes finding of the protocol buffers support when protobuf 1.2 is
present (#4)

ringerc committed:
Mention packages for Fedora

isaachier commented on .gitignore:
```diff
@@ -1 +1,4 @@
 /build
+DartConfiguration.tcl
+/CMakeFiles
+/CMakeCache.txt
```
For these two directories, please remove the leading slash. This pattern only matches `${SRC_ROOT}/CMakeFiles` but not `${SRC_ROOT}/build/CMakeFiles`. Usually I use `mkdir build && cd build && cmake ..` to keep the build in its own directory, so it made sense to ignore the whole build directory in the source root. However, it seems like you are building in the source tree. Might as well just ignore all of the CMake generated files no matter where they are.

isaachier replied:
You can actually replace this entirely with this: https://github.com/github/gitignore/blob/master/CMake.gitignore.

isaachier replied:
Maybe add this and any generated files you can find: https://github.com/github/gitignore/blob/master/C.gitignore.

isaachier commented on README.md:
````diff
@@ -10,4 +10,11 @@ sudo apt install libhttp-parser-dev libjansson-dev protobuf-c-compiler \
+sudo dnf install jansson-devel http-parser-devel protobuf-c-devel \
+		protobuf-c-compiler
+```
+
````
Nice! I haven't tested `HUNTER_DISABLED=ON` in a while, so I should double-check this still works.

ringerc commented:
@isaachier I've dropped the gitignore commit from the series, and I think otherwise it should be fine to merge.

ringerc commented:
Actually hold on this, needs more work first

ringerc committed:
Document cmake -DHUNTER_ENABLED=0 -DJAEGERTRACINGC_USE_PCG=0

ringerc committed:
Fix reference to a PCG header when PCG disabled

Amends 49513a5 (re #2)

ringerc commented:
I've removed the gitignore changes, added a fix to make it compile on protobuf-c 1.2, fixed some more issues with case in the CMake files, and rebased on top of 49513a53. It now configures fine here with `cmake -DHUNTER_ENABLED=0 -DJAEGERTRACINGC_USE_PCG=0` and builds fine.

ringerc committed:
Fixes for finding protobuf-c with !HUNTER_ENABLED

ringerc committed:
[travis] Quiet builds of protobuf and protobuf-c

ringerc committed:
[travis] Support building with the locally installed CMake on test host

ringerc committed:
[travis] Only build protobuf and protobuf-c when not using Hunter

ringerc committed:
[travis] When !HUNTER_ENABLED build and install opentracing in tests

ringerc committed:
[travis] Perform test builds with HUNTER_ENABLED=0

ringerc committed:
Disable some test matrix entries to reduce build load

ringerc committed:
Initialize system RNG when not using PCG

ringerc committed:
Expand README for local dependencies

isaachier commented on .travis.yml:
```diff
@@ -84,41 +101,43 @@ matrix:
-        - clang-4.0
-    env:
-    - MATRIX_EVAL="CC=clang-4.0 && CXX=clang++-4.0"
+#DISABLED to reduce the matrix a bit
```
I'm assuming this is for your own testing. In practice, we can add an additional matrix entry (probably one will suffice) that will test `HUNTER_ENABLED=0`.

isaachier commented on CMakeLists.txt:
```diff
@@ -113,19 +113,24 @@ if(HUNTER_ENABLED)
   find_package(Protobuf ${hunter_config} REQUIRED)
   list(APPEND package_deps Protobuf)
-  set(protoc_exe protobuf::protoc)
-else()
```
So how do we generate the messages? How do we connect to backend at all?

ringerc replied:
You will have seen the answer to that further down, guessing this is outdated. It's just a change in how the IDL is compiled by cmake.

isaachier commented on CMakeLists.txt:
```diff
@@ -113,19 +113,24 @@ if(HUNTER_ENABLED)
+  message(STATUS "using protoc and protoc-gen-c")
+  set(protoc_exe $<TARGET_FILE:protobuf::protoc>)
+  set(protoc_plugin_arg --plugin=$<TARGET_FILE:protobuf-c::protoc-gen-c>)
+elseif(TARGET protobuf-c::protoc-c)
```
I can fix this issue by using `protoc-c` directly. I found that I had issues doing this with the Hunter version of `protoc-c`, so I used an invocation of `protoc` that loaded the `protoc-gen-c` plugin. If Hunter is disabled, I can try using `protoc-c` directly instead.

ringerc replied:
That's what I did in an earlier revision, but there turns out to be no point. protobuf-c 1.3 adds `protobuf-gen-c` and is also the oldest version that supports "proto3". So in practice there's a hard requirement on 1.3 and you can just rely on doing it this way.

It's silly that Hunter doesn't expose a `protobuf-c::protoc-c` target, but in this case it wouldn't help us.

isaachier commented on README.md:
```diff
@@ -2,12 +2,72 @@
-Install the project library dependencies. On Debian, the following command
-should work.
+By default the client build will fetch all dependencies from the Hunter
+build dependency system. Only cmake must be installed.
```
:+1:

isaachier commented on README.md:
```diff
@@ -2,12 +2,72 @@
+
+On older Fedora releases, RHEL, and CentOS, protobuf-c-devel
+protobuf-c-compiler are too old and must be installed from source, custom
+packages, or a backport repository instead.
```
:) excellent!

isaachier commented on README.md:
```diff
@@ -2,12 +2,72 @@
+    git submodule update --init
+    mkdir -p build
+    cd build
+    cmake ..
```
Just minor suggestion: `cmake -DCMAKE_BUILD_TYPE=Release ..` or `Debug`.

isaachier commented on cmake/Findjansson.cmake:
```diff
@@ -55,9 +55,9 @@ if(JANSSON_FOUND)
-  set_target_properties(Jansson::Jansson PROPERTIES
+if(JANSSON_FOUND AND NOT TARGET jansson::jansson)
+  add_library(jansson::jansson UNKNOWN IMPORTED)
+  set_target_properties(jansson::jansson PROPERTIES
```
Fantastic! The casing can be annoying, but that seems to be the correct fix.

ringerc replied:
Very. CMake could at least offer a "hint" message of some kind. "Target Jansson::Jansson not found (but there's a jansson::jansson, did you mean that?"

isaachier commented on scripts/build.sh:
```diff
@@ -41,26 +41,52 @@ function main() {
+
+    if [ "${USE_HUNTER:-}" == "off" ]; then
+        # We won't be using Hunter so we need to get opentracing's c library
+        # installed too.
```
How would you prefer I handle this dependency to support non-Hunter builds?

ringerc replied:
I'm not sure yet, thinking about that. It depends a lot on whether you think apps would expect to link to `opentracing-c` then load `jaeger-client-c` at runtime, or if you see them linking to a `jaeger-client-c` that implements the `opentracing-c` APIs exposed in `opentracing-c` headers.

(The build doesn't produce a shared library yet. I'll deal with that at some later stage, far from urgent.)

You've pushed opentracing-c to hunter, so I guess you see it working more like the former? The app links to opentracing-c, and it loads a tracing implementation at runtime? I haven't looked closely at this part yet, as I've just been trying to get the build to work.

No strong opinions here anyway. I anticipate I'll be treating it as a unit of opentracing-c + jaeger-client-c anyway.

isaachier commented on scripts/install-deps.sh:
```diff
@@ -22,10 +22,10 @@ function build_from_tar() {
-    make -j3
-    make check -j3
-    make install
+    ./configure --quiet --prefix="$prefix"
```
Nice!

isaachier commented on src/jaegertracingc/random_test.c:
```diff
@@ -34,6 +38,14 @@ static void* random_func(void* arg)
+     * not much benefit when we're just generating request and span IDs, etc.
+     */
+    srand(time(NULL));
+#endif
```
Good idea.
